tests/autocomplete/test_autocomplete.py

```diff
@@ -377,3 +382,335 @@ def test_things_in_autocomplete_definition_file(
     assert actual_type == expected_type, (
         f"\n\nautocomplete_definition.py, line {line_number}: Expecting {expected_type} (but got {actual_type}) for the statement\n{statement}\n"
     )
+
+
+def test_all_table_methods():
+    all_methods_from_tables = [
+        f"{module.__name__}.{table_name}.{method}"
+        for module in get_submodules(tables)
+        for table_name, table in get_tables_from_namespace(module)
+        for method in table.__class__.__dict__
+        if callable(getattr(table, method)) and not method.startswith("_")
+    ]
+
+    # This is a list of table methods that we currently don't support for autocomplete,
+    # or need to ignore
+    ignored_table_methods = [
+        "ehrql.tables.smoketest.patients.count_for_patient",
+        "ehrql.tables.smoketest.patients.exists_for_patient",
+        "ehrql.tables.emis.patients.age_on",  # requires DateDifference
+        "ehrql.tables.tpp.addresses.for_patient_on",  # Needs last_for_patient
+        "ehrql.tables.core.practice_registrations.for_patient_on",  # Needs last_for_patient
+        "ehrql.tables.tpp.patients.age_on",  # requires DateDifference
+        "ehrql.tables.core.patients.age_on",  # requires DateDifference
+        "ehrql.tables.emis.ons_deaths.cause_of_death_is_in",  # need refactor of method
+        "ehrql.tables.core.ons_deaths.cause_of_death_is_in",  # need refactor of method
+    ]
+    tested_table_methods = [
+        f"ehrql.tables.{calling_object}.{method}"
+        for (calling_object, method) in autocomplete_file_methods
+    ]
+    missing_methods = [
+        method
+        for method in all_methods_from_tables
+        if method not in ignored_table_methods + tested_table_methods
+    ]
+    missing_methods_str = "\n  - ".join(missing_methods)
+    assert not missing_methods, (
+        "\nThe following methods do not have autocomplete tests:\n"
+        f"  - {missing_methods_str}\n"
+        "You must add them to the `autocomplete_definition.py` file.\n"
+        "If we don't support autocomplete (yet) then add them to the "
+        "`ignore_methods` list in this test file."
+    )
+
+
+@pytest.fixture(scope="session")
+def query_language_methods():
+    """
+    Fixture to extract all methods from the query_language.py file.
+    Determines for each method whether it is callable by a PatientSeries
+    or an EventSeries or both.
+
+    Currently there is one special case for the decorator @int_property.
+    This appears as a method in the below, but is actually an attribute
+    after the decorator has done its job. Therefore we deal with them
+    separately.
+    """
+    class_parents = {}
+    class_methods = {"object": []}
+    class_int_props = {"object": []}
+    series_methods = set()
+    series_int_props = set()
+    other_methods = {}
+    ql_classes = inspect.getmembers(query_language, inspect.isclass)
+    for class_name, class_obj in ql_classes:
+        if class_obj.__module__ == "ehrql.query_language":
```

Comment:
This might be slightly more readable written as a guard clause i.e.
```suggestion
        if class_obj.__module__ != "ehrql.query_language":
            continue
```
And then de-indenting the inner loop.

It  makes immediately clear the intention that you just want to ignore anything which isn't part of the query language module, which otherwise isn't obvious until you get to the end of the (quite long) conditional clause.

Reply:
Makes sense - thanks.